contrib/golang/filters/http/source/golang_filter.cc

```diff
@@ -913,6 +913,56 @@ CAPIStatus Filter::getStringValue(int id, GoString* value_str) {
   return CAPIStatus::CAPIOK;
 }
 
+CAPIStatus Filter::getDynamicMetadata(const std::string& filter_name, GoSlice* buf_slice) {
+  Thread::LockGuard lock(mutex_);
+  if (has_destroyed_) {
+    ENVOY_LOG(debug, "golang filter has been destroyed");
+    return CAPIStatus::CAPIFilterIsDestroy;
+  }
+
+  auto& state = getProcessorState();
+  if (!state.isProcessingInGo()) {
+    ENVOY_LOG(debug, "golang filter is not processing Go");
+    return CAPIStatus::CAPINotInGo;
+  }
+
+  auto dlib = Dso::DsoManager<Dso::HttpFilterDsoImpl>::getDsoByID(config_->soId());
+  ASSERT(dlib != nullptr, "load at the config parse phase, so it should not be null");
+  if (!state.isThreadSafe()) {
+    auto weak_ptr = weak_from_this();
+    ENVOY_LOG(debug, "golang filter getDynamicMetadata posting request to dispatcher");
+    state.getDispatcher().post([this, &state, weak_ptr, filter_name, buf_slice, dlib] {
+      ENVOY_LOG(debug, "golang filter getDynamicMetadata request in worker thread");
+      Thread::ReleasableLockGuard lock(mutex_);
+      if (!weak_ptr.expired() && !has_destroyed_) {
+        ASSERT(state.isThreadSafe());
+        populateSliceWithMetadata(state, filter_name, buf_slice);
+        dlib->envoyGoRequestSemaDec(req_);
+      } else {
+        ENVOY_LOG(info, "golang filter has gone or destroyed in getDynamicMetadata");
+      }
+    });
+  } else {
+    ENVOY_LOG(debug, "golang filter getDynamicMetadata replying directly");
+    populateSliceWithMetadata(state, filter_name, buf_slice);
+    dlib->envoyGoRequestSemaDec(req_);
```

Comment:
we could avoid this cgo call, when we introduce the `CAPI_YIELD` status for the `dispatcher.post` branch.
yep, it's just a performace optimization.
A c call Go call may take almost 1000 ns CPU time in the current Go implementation.
(we are optimizing it, may reduce to almost 100ns CPU time, it may be included in the comming Go 1.21).

Reply:
makes sense now